File: src/cli/headless.ts
import type { Agent } from '../agent.js';
import type { MessageParam, ContentBlock } from '../llm/anthropic.js';
import { sessionManager } from '../utils/session.js';

export async function runHeadless(
  agent: Agent,
  initialPrompt: string,
  sessionName?: string,
  resumeRecent?: boolean,
): Promise<void> {
  // Load session if requested
  if (sessionName || resumeRecent) {
    const name = sessionName ?? (await sessionManager.getMostRecent()) ?? `session-${Date.now()}`;
    const data = await sessionManager.get(name);
    if (data) {
      agent.setMessages(data.messages as any);
      const totalTokens = data.totalTokens || 0;
      if (totalTokens > 0) {
        agent.setTokenCount(totalTokens);
      }
      const { createLogger } = await import('../utils/logger.js');
      const newLogger = await createLogger(sessionManager.getProjectHash(), name);
      agent.setSession(name);
      agent.setLogger(newLogger);
    }
  }

  // Set headless display: confirm always denies
  agent.setDisplay({
    status: () => {},
    error: (msg) => console.error(`[error] ${msg}`),
    updateTokenCount: () => {},
    confirm: async (req) => {
      console.log(`[Permission denied: ${req.message}] -- use --permission yolo or auto in headless mode`);
      return false;
    },
  });

  let printedTurns = 0;
  const printedBlocks = new Map<number, number>();        // turnIndex → blocks printed so far
  const streamedChars = new Map<string, number>();         // "turnIdx:blockIdx" → chars printed
  const finalizedBlocks = new Set<string>();               // "turnIdx:blockIdx" → finalized (printed newline)
  const printedToolUses = new Set<string>();               // tool_use block IDs already printed
  const printedResults = new Set<string>();                // tool_use IDs whose results have been printed

  agent.getStore().onChange(() => {
    const turns = agent.getStore().getTurns();
    const statuses = agent.getStore().getStatuses();

    for (let ti = printedTurns; ti < turns.length; ti++) {
      const turn = turns[ti];
      if (turn.role === 'user') {
        if (typeof turn.content === 'string') {
          process.stdout.write(`[user] ${turn.content}\n\n`);
        }
        // tool_result turns are not printed separately; they appear under tool_use
        printedTurns = ti + 1;
        continue;
      }

      if (turn.role === 'assistant' && Array.isArray(turn.content)) {
        const blocks = turn.content as ContentBlock[];
        const blocksPrinted = printedBlocks.get(ti) || 0;

        for (let bi = blocksPrinted; bi < blocks.length; bi++) {
          const block = blocks[bi];
          const blockKey = `${ti}:${bi}`;

          // --- thinking: print when it stops growing ---
          if (block.type === 'thinking') {
            const prevLen = streamedChars.get(blockKey) || 0;
            if (block.thinking.length > prevLen) {
              // First time printing this thinking block
              if (prevLen === 0) process.stdout.write(`\n[thinking] `);
              process.stdout.write(block.thinking.slice(prevLen));
              streamedChars.set(blockKey, block.thinking.length);
            }
            // Check if this block is finalized (not the last block of the last turn)
            const isLastBlock = ti === turns.length - 1 && bi === blocks.length - 1;
            if (!isLastBlock && !finalizedBlocks.has(blockKey)) {
              process.stdout.write('\n');
              finalizedBlocks.add(blockKey);
            }
          }

          // --- text: stream incrementally ---
          if (block.type === 'text') {
            const prevLen = streamedChars.get(blockKey) || 0;
            if (block.text.length > prevLen) {
              if (prevLen === 0) process.stdout.write('\n[assistant] ');
              process.stdout.write(block.text.slice(prevLen));
              streamedChars.set(blockKey, block.text.length);
            }
            const isLastBlock = ti === turns.length - 1 && bi === blocks.length - 1;
            if (!isLastBlock && !finalizedBlocks.has(blockKey)) {
              process.stdout.write('\n');
              finalizedBlocks.add(blockKey);
            }
          }

          // --- tool_use: print call line + immediate result ---
          if (block.type === 'tool_use' && !printedToolUses.has(block.id)) {
            printedToolUses.add(block.id);
            const callText = `${block.name}(${JSON.stringify(block.input)})`;
            process.stdout.write(`\n[tool] ${callText}\n`);

            // Scan subsequent turns for matching tool_result
            for (let rti = ti + 1; rti < turns.length; rti++) {
              const rt = turns[rti];
              if (rt.role === 'user' && Array.isArray(rt.content)) {
                for (const rb of rt.content as any[]) {
                  if (rb.type === 'tool_result' && rb.tool_use_id === block.id) {
                    const raw = typeof rb.content === 'string' ? rb.content : JSON.stringify(rb.content);
                    const lines = raw.split('\n');
                    for (const line of lines) {
                      if (line) console.log(`       ${line}`);
                    }
                    printedResults.add(block.id);
                  }
                }
              }
            }
          }
        }

        // Track progress
        const isLastTurn = ti === turns.length - 1;
        if (!isLastTurn) {
          printedBlocks.set(ti, blocks.length);
          printedTurns = ti + 1;
        } else {
          // Don't finalize the last turn's last block — it may still be streaming
          printedBlocks.set(ti, blocks.length > 0 ? blocks.length - 1 : 0);
        }
      }
    }

    // Print tool results for any printed tool_use blocks
    for (const turn of turns) {
      if (turn.role === 'user' && Array.isArray(turn.content)) {
        for (const block of turn.content as any[]) {
          if (block.type === 'tool_result' && printedToolUses.has(block.tool_use_id) && !printedResults.has(block.tool_use_id)) {
            printedResults.add(block.tool_use_id);
            const raw = typeof block.content === 'string' ? block.content : JSON.stringify(block.content);
            const lines = raw.split('\n');
            for (const line of lines) {
              if (line) console.log(`       ${line}`);
            }
          }
        }
      }
    }

    // Print any new status messages
    for (const s of statuses) {
      if (s.role === 'error') console.error(`[error] ${s.content}`);
    }
  });

  try {
    await agent.run(initialPrompt);
  } catch (e) {
    if (e instanceof Error && e.message === 'Aborted') {
      console.log('(Aborted)');
    } else if (e instanceof Error) {
      console.error(`(Error: ${e.message})`);
    } else {
      throw e;
    }
  }
}

File: internal/headless.go
package internal

import (
	"context"
	"fmt"
	"os"
	"strings"
)

// RunHeadless executes the agent in headless (non-TUI) mode, streaming output to stdout.
func RunHeadless(ctx context.Context, ag *Agent, initialPrompt, displayPrompt string) error {
	// Track what has already been printed for incremental rendering
	printedTurns := 0
	printedBlocks := make(map[int]int)      // turnIndex → blocks printed
	streamedChars := make(map[string]int)   // "ti:bi" → chars printed
	finalizedBlocks := make(map[string]bool)
	printedToolUses := make(map[string]bool)
	printedResults := make(map[string]bool)

	render := func(isFinal bool) {
		turns := ag.store.Turns()

		for ti := printedTurns; ti < len(turns); ti++ {
			turn := turns[ti]
			isLastTurn := ti == len(turns)-1

			if turn.Role == "user" {
				content := ""
				if turn.Display != "" {
					content = turn.Display
				} else if s, ok := turn.Content.(string); ok {
					content = s
				}
				if content != "" {
					fmt.Fprintf(os.Stdout, "[user]\n%s\n\n", strings.TrimSpace(content))
				}
				printedTurns = ti + 1
				continue
			}

			if turn.Role != "assistant" {
				continue
			}

			var blocks []ContentBlock
			switch c := turn.Content.(type) {
			case []ContentBlock:
				blocks = c
			case []any:
				for _, raw := range c {
					if m, ok := raw.(map[string]any); ok {
						blocks = append(blocks, contentBlockFromMap(m))
					}
				}
			}

			bp := printedBlocks[ti]
			for bi := bp; bi < len(blocks); bi++ {
				block := blocks[bi]
				blockKey := fmt.Sprintf("%d:%d", ti, bi)
				isLastBlock := isLastTurn && bi == len(blocks)-1

				switch block.Type {
				case "thinking":
					prevLen := streamedChars[blockKey]
					if len(block.Thinking) > prevLen {
						if prevLen == 0 {
							fmt.Fprint(os.Stdout, "[thinking]\n")
						}
						content := block.Thinking
						if prevLen == 0 {
							content = strings.TrimLeft(content, " \t\n\r")
						} else {
							content = content[prevLen:]
						}
						fmt.Fprint(os.Stdout, content)
						streamedChars[blockKey] = len(block.Thinking)
					}
					if (!isLastBlock || isFinal) && !finalizedBlocks[blockKey] {
						if strings.HasSuffix(block.Thinking, "\n") {
							fmt.Fprint(os.Stdout, "\n")
						} else {
							fmt.Fprint(os.Stdout, "\n\n")
						}
						finalizedBlocks[blockKey] = true
					}
				case "text":
					prevLen := streamedChars[blockKey]
					if len(block.Text) > prevLen {
						if prevLen == 0 {
							fmt.Fprint(os.Stdout, "[assistant]\n")
						}
						content := block.Text
						if prevLen == 0 {
							content = strings.TrimLeft(content, " \t\n\r")
						} else {
							content = content[prevLen:]
						}
						fmt.Fprint(os.Stdout, content)
						streamedChars[blockKey] = len(block.Text)
					}
					if (!isLastBlock || isFinal) && !finalizedBlocks[blockKey] {
						if strings.HasSuffix(block.Text, "\n") {
							fmt.Fprint(os.Stdout, "\n")
						} else {
							fmt.Fprint(os.Stdout, "\n\n")
						}
						finalizedBlocks[blockKey] = true
					}
				case "tool_use":
					if !printedToolUses[block.ID] {
						printedToolUses[block.ID] = true
						fmt.Fprintf(os.Stdout, "[tool] %s(%s)\n", block.Name, jsonString(block.Input))

						// Scan subsequent turns for matching tool_result
						for rti := ti + 1; rti < len(turns); rti++ {
							rt := turns[rti]
							if rt.Role == "user" {
								switch rc := rt.Content.(type) {
								case []ContentBlock:
									for _, rb := range rc {
										if rb.Type == "tool_result" && rb.ToolUseID == block.ID {
											raw := rb.Content
											for _, line := range strings.Split(raw, "\n") {
												if line != "" {
													fmt.Fprintf(os.Stdout, "       %s\n", line)
												}
											}
											if !isLastBlock || isFinal {
												fmt.Fprintln(os.Stdout)
											}
											printedResults[block.ID] = true
										}
									}
								case []any:
									for _, raw := range rc {
										if rb, ok := raw.(map[string]any); ok {
											if rb["type"] == "tool_result" && rb["tool_use_id"] == block.ID {
												c, _ := rb["content"].(string)
												for _, line := range strings.Split(c, "\n") {
													if line != "" {
														fmt.Fprintf(os.Stdout, "       %s\n", line)
													}
												}
												if !isLastBlock || isFinal {
													fmt.Fprintln(os.Stdout)
												}
												printedResults[block.ID] = true
											}
										}
									}
								}
							}
						}
					}
				}
			}

			if !isLastTurn {
				printedBlocks[ti] = len(blocks)
				printedTurns = ti + 1
			} else if isFinal {
				printedBlocks[ti] = len(blocks)
				printedTurns = ti + 1
			} else {
				if len(blocks) > 0 {
					printedBlocks[ti] = len(blocks) - 1
				}
			}
		}

		// Print tool results for any printed tool_use blocks not yet shown
		for _, turn := range turns {
			if turn.Role == "user" {
				switch rc := turn.Content.(type) {
				case []ContentBlock:
					for _, block := range rc {
						if block.Type == "tool_result" && printedToolUses[block.ToolUseID] && !printedResults[block.ToolUseID] {
							printedResults[block.ToolUseID] = true
							for _, line := range strings.Split(block.Content, "\n") {
								if line != "" {
									fmt.Fprintf(os.Stdout, "       %s\n", line)
								}
							}
							fmt.Fprintln(os.Stdout)
						}
					}
				}
			}
		}

		// Print any new status/error messages
		printedStatus := make(map[int]bool)
		for _, s := range ag.Store().Statuses() {
			if !printedStatus[s.TurnIndex] {
				printedStatus[s.TurnIndex] = true
				if s.Role == RoleError {
					fmt.Fprintf(os.Stderr, "[error] %s\n", s.Content)
				}
			}
		}
	}

	ag.OnDisplayChange(func() { render(false) })

	ok, err := ag.Run(ctx, initialPrompt, displayPrompt)
	if err != nil {
		return err
	}
	_ = ok
	render(true)
	return nil
}

func jsonString(v any) string {
	if v == nil {
		return "{}"
	}
	switch val := v.(type) {
	case map[string]any:
		parts := make([]string, 0, len(val))
		for k, vv := range val {
			parts = append(parts, fmt.Sprintf("%q:%v", k, vv))
		}
		return "{" + strings.Join(parts, ", ") + "}"
	default:
		return fmt.Sprintf("%v", v)
	}
}
